### PR TITLE
Unify char checksum for archs with signed and unsigned char

### LINKED ISSUE #4583: FreeOrion fails to build on archs other that amd64
Bug Report
==========

Environment
-----------

* **FreeOrion Version**:  Weekly test build 2023-06-27.1b21f9b
* **Operating System**: Linux Ubuntu
* **Graphic card used**:
* **Fetched as** 
  * Compiled from source

Description
-----------

When trying to build PPA for Ubuntu I got compilation errors for different archs:

```
/<<PKGBUILDDIR>>/util/CheckSums.cpp:28:28: error: static assertion failed
   28 |     static_assert(char_cs2 == 61);
      |                   ~~~~~~~~~^~~~~
/<<PKGBUILDDIR>>/util/CheckSums.cpp:28:28: note: the comparison reduces to ‘(195 == 61)’
/<<PKGBUILDDIR>>/util/CheckSums.cpp:29:28: error: static assertion failed
   29 |     static_assert(char_cs3 == 15);
      |                   ~~~~~~~~~^~~~~
/<<PKGBUILDDIR>>/util/CheckSums.cpp:29:28: note: the comparison reduces to ‘(241 == 15)’
make[3]: *** [CMakeFiles/freeorioncommon.dir/build.make:779: CMakeFiles/freeorioncommon.dir/util/CheckSums.cpp.o] Error 1
```

https://launchpadlibrarian.net/674319701/buildlog_ubuntu-mantic-arm64.freeorion_0.5.999+1SNAPSHOT20230627ppa1~mantic_BUILDING.txt.gz

Expected Result
---------------

Successful build.

Steps to reproduce
------------------

* Build FreeOrion for Linux on non-amd64 arch.

### PR DESCRIPTION
**Pull request by o01eg**
Should fix #4583

### CODE REVIEW

File: util/CheckSums.h [hunk 1]
```diff
@@ -19,6 +19,14 @@ namespace CheckSums {
     constexpr bool csc_float = noexcept(noexcept(std::log10(std::abs(43.0f))));
     FO_COMMON_API void CheckSumCombine(uint32_t& sum, float t) noexcept(csc_float);
 
+    // unify char checksum for archs with singed and unsigned char
+    constexpr void CheckSumCombine(uint32_t& sum, char c)
+    {
+        signed char t = static_cast<signed char>(c);
```
Comment (geoffthemedio) [inline util/CheckSums.h:25]: doesn't produce consistent results for `signed char` and `unsigned char` constructed with values outside the range 0-127, eg:
````
    static constexpr auto char_cs2a = []() {
        uint32_t sum = 0;
        constexpr auto num = static_cast<signed char>(195);
        CheckSumCombine(sum, num);
        return sum;
    }();
    static constexpr auto char_cs2b = []() {
        uint32_t sum = 0;
        constexpr auto num = static_cast<unsigned char>(195);
        CheckSumCombine(sum, num);
        return sum;
    }();

    static constexpr auto char_cs3a = []() {
        uint32_t sum = 0;
        constexpr auto num = static_cast<signed char>(-15);
        CheckSumCombine(sum, num);
        return sum;
    }();
    static constexpr auto char_cs3b = []() {
        uint32_t sum = 0;
        constexpr auto num = static_cast<unsigned char>(-15);
        CheckSumCombine(sum, num);
        return sum;
    }();

    static_assert(char_cs2a == 61);
    static_assert(char_cs2b == 61);
    static_assert(char_cs3a == 15);
    static_assert(char_cs3b == 15);
````

Comment (o01eg) [reply util/CheckSums.h:25]: Maybe it because we change sign but instead should use 2's complement for consistency?

Comment (hifron) [reply util/CheckSums.h:25]: Has C++ chars somehow knowledge about UTF8, UTF16 and UTF32(maybe in some Asian langs or special chars like smiles or similar glyphs)?

Comment (o01eg) [reply util/CheckSums.h:25]: No, there other types for Unicode characters in C++.

**Comment by o01eg** [comment]
Superseded by #4585

**Pull request closed without merging**